api/v1/standalone-dns-proxy/standalone-dns-proxy.proto

```diff
@@ -0,0 +1,85 @@
+  // In case of any server side error, cilium agent will send an error response and SDP will have to re-subscribe.  
+  rpc StreamPolicyState(stream PolicyStateResponse) returns (stream PolicyState) {}
+  
+  // UpdateMappingRequest is used by the Standalone DNS proxy to update ciliium agent with FQDN-IP mappings which in turn update L3/L4 policy maps.
+  // In case of any error, SDP will either retry the connection if the error is server side or will error out.
+  // Note: In case of concurrent updates, since this is called in a callback(notifyDNSMsg) from the DNS server it follows the same behavior as
+  // the inbuilt dns proxy in cilium.
+  rpc UpdateMappingRequest(FQDNMapping) returns (UpdateMappingResponse) {}
+}
+
+// Response code returned by RPC methods.
+enum ResponseCode {
+    RESPONSE_CODE_UNSPECIFIED = 0;
+    RESPONSE_CODE_NO_ERROR = 1;
+    RESPONSE_CODE_FORMAT_ERROR = 2;
+    RESPONSE_CODE_SERVER_FAILURE = 3;
+    RESPONSE_CODE_NOT_IMPLEMENTED = 4;
+    RESPONSE_CODE_REFUSED = 5;
+}
+
+// Ack sent from SDP to Agent on processing DNS policy rules
+message PolicyStateResponse  {
+    ResponseCode response = 1;
+    string request_id = 2; // Request ID for which response is sent to
+}
+
+// FQDN-IP mapping goalstate sent from SDP to agent 
+message FQDNMapping {
+    string fqdn = 1; // dns name
+    repeated bytes record_ip = 2; // List of IPs corresponding to dns name
+    uint32 ttl = 3;  // TTL of DNS record
+    uint32 source_identity = 4; // Identity of the client making the DNS request
+    bytes source_ip = 5; // IP address of the client making the DNS request
+    uint32 response_code = 6; // DNS Response code as specified in RFC2316
+}
+
+// Ack returned by cilium agent to SDP on receiving FQDN-IP mapping update
+message UpdateMappingResponse {
+    ResponseCode response = 1;
+}
+
+// DNServer identity, port and protocol the requests be allowed to
+message DNSServer {
+    uint32 dns_server_identity = 1;  // Identity of destination DNS server
+    uint32 dns_server_port = 2;
+    uint32 dns_server_proto = 3;
+}
+
+// L7 DNS policy specifying which requests are permitted to which DNS server
+message DNSPolicy {
+    uint32 source_identity = 1;  // Identity of the workload this L7 DNS policy should apply to
+    repeated string dns_pattern = 2;  // Allowed DNS pattern this identity is allowed to resolve.
+    repeated DNSServer dns_servers = 3; // List of DNS servers to be allowed to
+}
+
+// L7 DNS policy snapshot of all local endpoints and identity to ip mapping of source 
+// and destinatione egress endpoints enforcing fqdn rules.
+message PolicyState {
+    repeated DNSPolicy egress_l7_dns_policy = 1;
+    string request_id = 2; // Random UUID based identifier which will be referenced in ACKs
+    repeated IdentityToIPMapping identity_to_ip_mapping = 3; // Identity to IP mapping for the DNS server and the source identity
```

Comment:
I don't quite understand how this works just yet. Is the plan to send a stream of these messages containing either DNS policy or identity/IP mappings? Or both?

Say the agent allocates a new identity, how does a message look like which transports this information to the proxy?

Reply:
it will send both on any update.

> Say the agent allocates a new identity, how does a message look like which transports this information to the proxy?

we added a listener callback on ipcache updates and that callback will retrieve policy snapshot and update ip-identity map and send update to SDP.

Reply:
Why don't we send differential updates? It doesn't seem necessary nor like it would scale to send the whole ip/identity map.

Reply:
just to clarify, we are sending ip->identity mappings only for endpoints that enforce fqdn(not for all). If we send only differential update, then SDP needs some way to detect if entry is deleted or updated which can be done by adding another field to indicate update/delete. The main concern is if sdp didn't receive / process update for some reason, then we need to worry about retries,etc and it may not recover until restarted. While we send snapshot of ip-identity mapping, even if first snapshot missed, then successive snapshots helps it to recover. sending snapshot of ip->identity mapping only for local endpoints which enforce fqdn is not that expensive

Reply:
I have a feeling this will not scale well for large clusters, but maybe that's not what we are aiming for. I see the concerns regarding retries, but there are guarantees that communicating locally (there's no network in between the agent and the proxy) over gRPC give you - but you would have to think about correctly implementing something like ListAndWatch in k8s.
Let's continue with this for now, then.